### PR TITLE
Support license type for Azure Hybrid Benefit

### LINKED ISSUE #154: Packer Azure builder doesn't support choosing license
_This issue was originally opened by @maciejsikora2302 in https://github.com/hashicorp/packer/issues/11278 and has been migrated to this repository. The original issue description is below._

---
There is no way in packer to choose with which license a VM should be created.

[Official Azure guide to hybrid licensing](https://docs.microsoft.com/en-us/azure/virtual-machines/windows/hybrid-use-benefit-licensing)

### PR DESCRIPTION
**Pull request by dknott-arcadis**
Adds support for specifying the license type for the builder virtual machine.

- Supports all 4 (as of this PR) Hybrid Benefit modes:
  - [RHEL](https://learn.microsoft.com/en-us/azure/virtual-machines/linux/azure-hybrid-benefit-linux)
  - [SUSE Enterprise](https://learn.microsoft.com/en-us/azure/virtual-machines/linux/azure-hybrid-benefit-linux)
  - [Windows Server](https://learn.microsoft.com/en-us/azure/virtual-machines/windows/hybrid-use-benefit-licensing)
  - [Windows Client](https://learn.microsoft.com/en-us/azure/virtual-machines/windows/hybrid-use-benefit-licensing)
- Normalizes the case of the configured value to match [Microsoft's API documentation](https://learn.microsoft.com/en-us/azure/templates/Microsoft.Compute/2021-03-01/virtualmachines?pivots=deployment-language-arm-template).
- Validates the value to ensure that it is valid in combination with the specified OS Type
- Tests added to cover the new functionality.

Closes #154

### CODE REVIEW

File: builder/azure/arm/config.go [hunk 1]
```diff
@@ -1240,6 +1259,30 @@ func assertRequiredParametersSet(c *Config, errs *packersdk.MultiError) {
...
+		if c.OSType == constants.Target_Linux {
+			if strings.EqualFold(c.LicenseType, constants.License_RHEL) {
+				c.LicenseType = constants.License_RHEL
+			} else if strings.EqualFold(c.LicenseType, constants.License_SUSE) {
+				c.LicenseType = constants.License_SUSE
+			} else {
+				errs = packersdk.MultiErrorAppend(errs, fmt.Errorf("The license_type %q is invalid for os_type %q", c.LicenseType, c.OSType))
+			}
+		} else if c.OSType == constants.Target_Windows {
+			if strings.EqualFold(c.LicenseType, constants.License_Windows_Client) {
```
Comment (agrawalkshitij) [inline builder/azure/arm/config.go:1277]: you are just trying to make sure the license type is of certain valid values, so this code can be simplified to 
if(!strings.EqualFold(c.LicenseType, constants.License_Windows_Client) &&
   !strings.EqualFold(c.LicenseType, constants.License_Windows_Server))
{
errs = packersdk.MultiErrorAppend(errs, fmt.Errorf("The license_type %q is invalid for os_type %q", c.LicenseType, c.OSType))
}

**Review by agrawalkshitij — APPROVED** [review]
Looks good to me

File: builder/azure/arm/config.go (hunk 1 above)
Comment (dknott-arcadis) [reply builder/azure/arm/config.go:1277]: It's also normalizing the case of LicenseType to avoid any case sensitivity issues with the Azure API.

Comment (agrawalkshitij) [reply builder/azure/arm/config.go:1277]: makes sense!

File: docs-partials/builder/azure/arm/Config-not-required.mdx [hunk 2]
```diff
@@ -325,6 +325,22 @@
   this will set the prefix for the resources. The actuall resource names will be
   `custom_resource_build_prefix` + resourcetype + 5 character random alphanumeric string
 
+- `license_type` (string) - Specify a license type for the VM to enable Azure Hybrid Benefit. If not set, Pay-As-You-Go license
```
Comment (JenGoldstrich) [inline docs-partials/builder/azure/arm/Config-not-required.mdx:328]: Very nice

File: builder/azure/arm/config.go [hunk 3]
```diff
@@ -1240,6 +1259,30 @@ func assertRequiredParametersSet(c *Config, errs *packersdk.MultiError) {
...
+				c.LicenseType = constants.License_Windows_Client
+			} else if strings.EqualFold(c.LicenseType, constants.License_Windows_Server) {
+				c.LicenseType = constants.License_Windows_Server
+			} else {
+				errs = packersdk.MultiErrorAppend(errs, fmt.Errorf("The license_type %q is invalid for os_type %q", c.LicenseType, c.OSType))
```
Comment (JenGoldstrich) [inline builder/azure/arm/config.go:1282]: I think this error message can be a bit more specific

```suggestion
				errs = packersdk.MultiErrorAppend(errs, fmt.Errorf("The license_type %q is invalid for Windows, use Windows_Client or Windows_Server", c.LicenseType))
```

File: builder/azure/arm/config.go [hunk 4]
```diff
@@ -1240,6 +1259,30 @@ func assertRequiredParametersSet(c *Config, errs *packersdk.MultiError) {
...
+	if c.LicenseType != "" {
+		// Assumes OS is case-sensitive match as it has already been
+		// normalized earlier in function
+		if c.OSType == constants.Target_Linux {
+			if strings.EqualFold(c.LicenseType, constants.License_RHEL) {
+				c.LicenseType = constants.License_RHEL
+			} else if strings.EqualFold(c.LicenseType, constants.License_SUSE) {
+				c.LicenseType = constants.License_SUSE
+			} else {
+				errs = packersdk.MultiErrorAppend(errs, fmt.Errorf("The license_type %q is invalid for os_type %q", c.LicenseType, c.OSType))
```
Comment (JenGoldstrich) [inline builder/azure/arm/config.go:1274]:
```suggestion
				errs = packersdk.MultiErrorAppend(errs, fmt.Errorf("The license_type %q is invalid for Linux, only RHEL_BYOS or SLES_BYOS are supported", c.LicenseType))
```

**Review by JenGoldstrich — COMMENTED** [review]
Code is looking good so far, made some suggested changes to the error messaging let me know what you think!

**Comment by dknott-arcadis** [comment]
> Code is looking good so far, made some suggested changes to the error messaging let me know what you think!

Good suggestions, I applied them. 😄

**Comment by agrigorev1337** [comment]
Hi guys, hope you could find some time to complete the PR. A lot of teams are blocked right now because of this issue. We will really appreciate your help here!

File: docs-partials/builder/azure/arm/Config-not-required.mdx [hunk 5]
```diff
@@ -327,6 +327,22 @@
...
```
Comment (JenGoldstrich) [inline docs-partials/builder/azure/arm/Config-not-required.mdx:330]:
```suggestion
- `license_type` (string) - Specify a license type for the build VM to enable Azure Hybrid Benefit. If not set, Pay-As-You-Go license.
```

**Review by JenGoldstrich — APPROVED** [review]
I was able to manually verify this, and run acceptance tests, worth noting the Packer team doesn't have an active account to test Red Hat/SUSE, but logically I think they should work, I tested Windows Server and Windows Client.

Thanks for this contribution @dknott-arcadis!

**Pull request merged**